### PR TITLE
Exclude added flash types from action_methods

### LINKED ISSUE #33593: Added flash types are listed in controller action methods
### Steps to reproduce

1. Subclass `ActionController::Base`
2. Add an additional flash type via `add_flash_types`
3. Inspect that controller's `action_methods`
4. You'll see your flash type included in that set

``` ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
  end
end

class TestController < ActionController::Base
  add_flash_types :hype
end

require "minitest/autorun"
require "rack/test"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_returns_success
    assert TestController.action_methods.include?('hype')
  end
end

```

### Expected behavior

Flash types should not be listed as action methods.

### Actual behavior
Flash types are indeed being listed as action methods.

### System configuration
**Rails version**:
Observed in 5.2.1 and edge.

**Ruby version**:
2.5.1

### PR DESCRIPTION
**Pull request by Bodacious**
Exclude additional flash types from `action_methods`.

Fixes #33593

When new flash types are added to ApplicationController, these
should not appear in the `action_methods` set, since they're
not actions. This commit corrects this behaviour.

    class MyController < ApplicationController
      add_flash_types :hype
    end

    # Currently:
    MyController.action_methods.include?('hype') # => true

    # This commit
    MyController.action_methods.include?('hype') # => false

### CODE REVIEW

**Comment by schneems** [comment]
Tests are still failing. Can you take a look?

**Comment by Bodacious** [comment]
Yes, will do today!   
  

  
Cheers.
  

  
  

  
  
>

**Comment by Bodacious** [comment]
@schneems I would appreciate your input on the approach here.

On closer inspection, the errors that were breaking tests were caused by `ActionMailer::Base`'s definition of `respond_to_missing?`.

My initial changes to `action_methods` did a quick `respond_to?` check to make sure the object responds to `_flash_types`. Since this isn't defined in ActionMailer::Base, this would call `respond_to_missing?`. And since `respond_to_missing?` calls `action_methods` this created an infinite loop.

My solution here was to add an additional method to store the flash types, one that will be empty in the case of an `ActionMailer::Base` object. But this is extending that class's interface with a method it doesn't require (violation of the Interface segregation principle?).

An alternative solution might be to extend `ActionMailer::Base#respond_to_missing?` to check if the missing method being called is `_flash_types`. But it didn't seem right modifying a very general method for a very particular purpose.

Another alternative still might be to introspect on the list of instance methods defined on the controller, instead of using `respond_to?`. Something like: `methods.include?(:_flash_types)`. But this didn't seem very elegant.

Any input would be appreciated. Thanks!

**Comment by oniofchaos** [comment]
Tests look like they are passing now? Just need to fix merge conflicts.

**Comment by rafaelfranca** [comment]
Can't we fix this problem by defining the flash methods as private?

**Comment by Bodacious** [comment]
That would be a much nearer solution, but would break existing code that calls those methods on the controller.

**Comment by oniofchaos** [comment]
Looks like there's a merge conflict on the changelog now 😢

**Comment by Bodacious** [comment]
@oniofchaos I've resolved the new conflicts. Would you mind having a look at this PR soon, so I don't have another merge conflict 😩

File: actionpack/lib/abstract_controller/base.rb [hunk 1]
```diff
@@ -73,19 +73,28 @@ def internal_methods
...
           methods = (public_instance_methods(true) -
             # Except for public instance methods of Base and its ancestors
             internal_methods +
             # Be sure to include shadowed public instance methods of this class
-            public_instance_methods(false))
-
-          methods.map!(&:to_s)
+            public_instance_methods(false) -
+            # Except for flash methods added to this controller class
+            added_flash_types.to_a).map(&:to_s).uniq
```
Comment (yskkin) [inline actionpack/lib/abstract_controller/base.rb]: Do we need both of `uniq` here and `to_set` below?

Comment (Bodacious) [reply actionpack/lib/abstract_controller/base.rb]: Well caught!

**Comment by yskkin** [comment]
`is_a?(ActionController::Flash) ? _flash_types : []` can be another option.
If `AbstractController::Base` should not know about flash, we can override `action_methods` in `ActionController::Base`.

**Comment by Bodacious** [comment]
@yskkin Thank you for your feedback. This new solution is simpler and cleaner.

File: actionpack/test/controller/flash_test.rb [hunk 2]
```diff
@@ -224,6 +224,16 @@ def test_redirect_to_with_adding_flash_types
...
   end
 
+  def test_additional_flash_types_are_not_listed_in_actions_set
+    original_controller = @controller
+    test_controller_with_flash_type_foo = Class.new(TestController) do
+      add_flash_types :foo
+    end
+    assert_not_includes test_controller_with_flash_type_foo.action_methods, "foo"
+  ensure
+    @controller = original_controller
```
Comment (yskkin) [inline actionpack/test/controller/flash_test.rb]: I don't think we need to backup and restore `@controller` since it is never used in this test case.

**Comment by Bodacious** [comment]
@yskkin Thank you! I've corrected that

**Comment by Bodacious** [comment]
@schneems Any chance someone could look at this PR again soon? Thanks.

**Pull request closed without merging**